src/main/java/frc/robot/subsystems/swerve/PhoenixOdometryThread.java

```diff
@@ -51,88 +59,97 @@ public static PhoenixOdometryThread getInstance() {
-    signalsLock.lock();
-    SwerveSubsystem.odometryLock.lock();
+  public void registerSignals(Collection<Registration> registrations) {
+    registerSignals(registrations.toArray(new Registration[] {}));
+  }
+
+  // Returns a handle which can be used to collect the last 20 signal results
+  public void registerSignals(Registration... registrations) {
+    var writeLock = journalLock.writeLock();
+
     try {
-      isCANFD = CANBus.isNetworkFD(device.getNetwork());
-      BaseStatusSignal[] newSignals = new BaseStatusSignal[signals.length + 1];
-      System.arraycopy(signals, 0, newSignals, 0, signals.length);
-      newSignals[signals.length] = signal;
-      signals = newSignals;
-      queues.add(queue);
+      writeLock.lock();
+
+      for (var registration : registrations) {
+        assert CANBus.isNetworkFD(registration.device.getNetwork()) : "Only CAN FDs supported";
+
+        signals.addAll(registration.signals);
+      }
     } finally {
-      signalsLock.unlock();
-      SwerveSubsystem.odometryLock.unlock();
+      writeLock.unlock();
     }
-    return queue;
   }
 
-  public Queue<Double> makeTimestampQueue() {
-    Queue<Double> queue = new ArrayDeque<>(100);
-    SwerveSubsystem.odometryLock.lock();
+  public List<Samples> samplesSince(double timestamp, Set<StatusSignal<Double>> signals) {
+    var readLock = journalLock.readLock();
     try {
-      timestampQueues.add(queue);
+      readLock.lock();
+
+      return journal.stream()
+          .filter(s -> s.timestamp > timestamp)
+          .map(
+              s -> {
+                var filteredValues =
+                    s.values.entrySet().stream()
+                        .filter(e -> signals.contains(e.getKey()))
+                        .collect(
+                            Collectors.toUnmodifiableMap(Map.Entry::getKey, Map.Entry::getValue));
+                return new Samples(s.timestamp, filteredValues);
+              })
+          .collect(Collectors.toUnmodifiableList());
     } finally {
-      SwerveSubsystem.odometryLock.unlock();
+      readLock.unlock();
     }
-    return queue;
   }
 
   @Override
   public void run() {
     while (true) {
       // Wait for updates from all signals
-      signalsLock.lock();
-      try {
-        if (isCANFD) {
-          BaseStatusSignal.waitForAll(2.0 / Module.ODOMETRY_FREQUENCY_HZ, signals);
-        } else {
-          // "waitForAll" does not support blocking on multiple
-          // signals with a bus that is not CAN FD, regardless
-          // of Pro licensing. No reasoning for this behavior
-          // is provided by the documentation.
-          Thread.sleep((long) (1000.0 / Module.ODOMETRY_FREQUENCY_HZ));
-          if (signals.length > 0) BaseStatusSignal.refreshAll(signals);
-        }
-      } catch (InterruptedException e) {
-        e.printStackTrace();
-      } finally {
-        signalsLock.unlock();
-      }
+      var writeLock = journalLock.writeLock();
```

Comment:
is there any specific reason why this isn't the same writeLock as before (just wondering since the old code had the same signalsLock)

Reply:
A read/write lock is a kind of mutual exclusion construct that allows multiple readers, but only one writer. That means if we ever used this in an environment with more than one thread calling `samplesSince`, multiple callers could coexist in between journal writes. Once a write begins, all the readers would have to wait, but once the write was done they'd be able to coexist without issue. The other lock allows only one access at a time, regardless of what's going on.

Reply:
In our current use it really doesn't make a difference. But it's a minor change that makes the class more future proof.